connect-inject/endpoints_controller.go

```diff
@@ -0,0 +1,449 @@
+package connectinject
+
+import (
+	"context"
+	"fmt"
+	"strings"
+
+	mapset "github.com/deckarep/golang-set"
+	"github.com/go-logr/logr"
+	"github.com/hashicorp/consul-k8s/consul"
+	"github.com/hashicorp/consul/api"
+	corev1 "k8s.io/api/core/v1"
+	k8serrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/types"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+type EndpointsController struct {
+	client.Client
+	// ConsulClient points at the agent local to the connect-inject deployment pod
+	ConsulClient *api.Client
+	// ConsulScheme is the scheme to use when making API calls to Consul,
+	// i.e. "http" or "https".
+	ConsulScheme string
+	// ConsulPort is the port to make HTTP API calls to Consul agents on.
+	ConsulPort string
+	// Only endpoints in the AllowK8sNamespacesSet are reconciled.
+	AllowK8sNamespacesSet mapset.Set
+	// Endpoints in the DenyK8sNamespacesSet are ignored.
+	DenyK8sNamespacesSet mapset.Set
+	// ReleaseName is the Consul Helm installation release.
+	ReleaseName string
+	// ReleaseNamespace is the namespace where Consul is Helm installed.
+	ReleaseNamespace string
+	Log              logr.Logger
+	Scheme           *runtime.Scheme
+	Ctx              context.Context
+}
+
+const MetaKeyKubeServiceName = "k8s-service-name"
+
+func (r *EndpointsController) Reconcile(req ctrl.Request) (ctrl.Result, error) {
+	var serviceEndpoints corev1.Endpoints
+
+	if shouldIgnore(req.Namespace, r.DenyK8sNamespacesSet, r.AllowK8sNamespacesSet) {
+		return ctrl.Result{}, nil
+	}
+
+	err := r.Client.Get(r.Ctx, req.NamespacedName, &serviceEndpoints)
+
+	// If the endpoints object has been deleted (and we get an IsNotFound
+	// error), we need to deregister all instances in Consul for that service.
+	if k8serrors.IsNotFound(err) {
+		// Deregister all instances in Consul for this service. The function deregisterServiceOnAllAgents handles
+		// the case where the Consul service name is different from the K8s service name.
+		err := r.deregisterServiceOnAllAgents(req.Name, req.Namespace)
+		if err != nil {
+			return ctrl.Result{}, err
+		}
+		return ctrl.Result{}, nil
+	} else if err != nil {
+		r.Log.Error(err, "failed to get endpoints from Kubernetes", "namespace", req.Namespace, "name", req.Name)
+		return ctrl.Result{}, err
+	}
+
+	r.Log.Info("retrieved service from kube", "serviceEndpoints", serviceEndpoints)
```

Comment:
The whole serviceEndpoints object shouldn't be logged here, just the name, same for other logs. Make sure to decide what should be Debug vs Info. Should info logs log every reconcile or only debug?